### PR TITLE
Protection against CSRF added

### PR DESCRIPTION
**Pull request by pboguslawski**
Details sent to security@gitea.io in messages

Wed, 30 Nov 2022 12:47:33 +0100.
Fri, 2 Dec 2022 19:59:02 +0100

Author-Change-Id: IB#1129006

### CODE REVIEW

File: routers/common/middleware.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021 The Gitea Authors. All rights reserved.
+// Copyright 2022 The Gitea Authors. All rights reserved.
```
Comment (mscherer) [inline routers/common/middleware.go]: AFAIK, updating the copyright year is kinda useless nowadays, as using git is sufficient to establish who/when the work was done (which is why there is "the gitea authors" and not the detail). That's the opinion of the Linux Foundation ( https://www.linuxfoundation.org/blog/blog/copyright-notices-in-open-source-software-projects , the author is the Vice President of Compliance and Legal ). I also got the same answer from our legal department, even if I haven't been able to convince them to publish something outside our intranet.

File: routers/common/middleware.go [hunk 2]
```diff
@@ -78,5 +78,43 @@ func Middlewares() []func(http.Handler) http.Handler {
...
+			// Put header names we use for CSRF recognition into Vary response header.
+			if setting.CORSConfig.Enabled {
+				resp.Header().Set("Vary", "Origin, Sec-Fetch-Site")
+			} else {
+				resp.Header().Set("Vary", "Sec-Fetch-Site")
+			}
+
+			// Allow requests not recognized as CSRF.
+			secFetchSite := strings.ToLower(req.Header.Get("Sec-Fetch-Site"))
+			if req.Method == "GET" || // GET must not be used for changing state (CSRF resistant).
```
Comment (mscherer) [inline routers/common/middleware.go]: Shouldn't it also cover HEAD ?

**Comment by zeripath** [comment]
I actually have a feeling that we should not be allowing the ReverseProxy Auth on the API - it's the only consistent way of making the API work

**Comment by zeripath** [comment]
I don't fully understand why you have added another dependency library for the cors instead of making our current one work or replacing our current one.

File: routers/common/middleware.go (hunk 1 above)
Comment (pboguslawski) [reply routers/common/middleware.go]: Fixed.

File: routers/common/middleware.go (hunk 2 above)
Comment (pboguslawski) [reply routers/common/middleware.go]: Added HEAD and OPTIONS also.

**Comment by pboguslawski** [comment]
> it's the only consistent way of making the API work

What inconsistency you see in using one auth method for all API calls?

> I don't fully understand why you have added another dependency library for the cors instead of making our current one work or replacing our current one.

https://github.com/go-chi/cors [does not provide](https://github.com/go-chi/cors/issues/21) `OriginAllowed` method and [it looks like](https://github.com/go-chi/cors/issues/21#issuecomment-1250908869) even maintainer is not sure why this fork is still requred for. Existing CSRF code was not analyzed and left untouched; if this PR is accepted - existing token based CSRF stuff should be checked and (if is still required as separate line of defense) may be switched to github.com/rs/cors or removed (if not required when browsers compatible with fetch metadata headers will be required for gitea).

**Comment by pboguslawski** [comment]
> For every other authentication mechanism we now require that API calls are separately and explicitly authenticated - this makes ReverseProxy the odd one out.

If account A can make change X using "web API" why shouldn't be allowed to make change X using "API"? Why to "mess" with separate API areas and separate sets of auths for both? Any real risks?

If separation is required - why not to introduce user permissions to access given API area (i.e. "[x] access to web", "[x] access to API) as alternative to application tokens (which should be optional)?

> This PR means we now have TWO libraries adding CORS headers. 
Code in this PR does not handle CORS requests it only uses CORS stuff to decide if some requests are valid. If this PR is accepted - existing CORS may be switched to github.com/rs/cors  to avoid using two CORS libs (should not hurt as temporary solution).

**Comment by zeripath** [comment]
> If account A can make change X using "web API" why shouldn't be allowed to make same change X using "API"? Why to "mess" with separate API areas and separate sets of auths for both? Any real risks?

I didn't complete agree with the decision to do this - but it was done. ReverseProxy currently stands as the odd one out. 

The idea is that you have to authenticate explicitly for the API separately from the UI.
 
> If separation is required - why not to introduce user permissions to access given API area (i.e. "[x] access to web", "[x] access to API) as alternative to application tokens (which should be optional)?

Or use tokens that we already have.

> > This PR means we now have TWO libraries adding CORS headers.
> 
> Code in this PR does not handle CORS requests/reponses it only uses CORS stuff to decide if requests are valid during CSRF validation. If this PR is accepted - existing CORS may be switched to github.com/rs/cors to avoid using two CORS libs (should not hurt as temporary solution).

That's not the way to do things, it won't get done and we'll end up with 2 slightly different CORS libraries. The go-chi/cors library should be replaced in this PR. I won't approve this PR without that although I'm not going to block it.

**Pull request closed without merging**